src/main/java9/org/sqlite/nativeimage/SqliteJdbcFeature.java

```diff
@@ -0,0 +1,157 @@
+package org.sqlite.nativeimage;
+
+import org.graalvm.nativeimage.hosted.Feature;
+import org.graalvm.nativeimage.hosted.RuntimeClassInitialization;
+import org.graalvm.nativeimage.hosted.RuntimeJNIAccess;
+import org.graalvm.nativeimage.hosted.RuntimeResourceAccess;
+import org.sqlite.*;
+import org.sqlite.core.DB;
+import org.sqlite.core.NativeDB;
+import org.sqlite.jdbc3.JDBC3DatabaseMetaData;
+import org.sqlite.util.LibraryLoaderUtil;
+import org.sqlite.util.OSInfo;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.lang.reflect.Field;
+import java.lang.reflect.Method;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.nio.file.StandardCopyOption;
+
+public class SqliteJdbcFeature implements Feature {
+
+    @Override
+    public void beforeAnalysis(BeforeAnalysisAccess a) {
+        RuntimeClassInitialization.initializeAtBuildTime(SQLiteJDBCLoader.VersionHolder.class);
+        RuntimeClassInitialization.initializeAtBuildTime(JDBC3DatabaseMetaData.class);
+        RuntimeClassInitialization.initializeAtBuildTime(OSInfo.class);
+        RuntimeClassInitialization.initializeAtBuildTime(LibraryLoaderUtil.class);
+        a.registerReachabilityHandler(
+                this::nativeDbReachable, method(SQLiteJDBCLoader.class, "initialize"));
+    }
+
+    private void nativeDbReachable(DuringAnalysisAccess a) {
+        handleLibraryResources();
+        registerJNICalls();
+    }
+
+    private void handleLibraryResources() {
+        String libraryPath = LibraryLoaderUtil.getNativeLibResourcePath();
+        String libraryName = LibraryLoaderUtil.getNativeLibName();
+        // Sanity check
+        if (!LibraryLoaderUtil.hasNativeLib(libraryPath, libraryName)) {
+            throw new SqliteJdbcFeatureException(
+                    "Unable to locate the required native resources for native-image. Please contact the maintainers of sqlite-jdbc.");
+        }
+
+        // libraryResource always has a leading '/'
+        String libraryResource = libraryPath + "/" + libraryName;
+        String exportLocation = System.getProperty("org.sqlite.lib.exportPath", "");
+        if (exportLocation.isEmpty()) {
+            // Do not export the library and include it in native-image instead
+            RuntimeResourceAccess.addResource(
+                    SQLiteJDBCLoader.class.getModule(), libraryResource.substring(1));
+        } else {
```

Comment:
Currently only this case is tested.
I documented the other case as "recommended" use-case, but most people will probably still just use the as-is setup without bothering with additional configuration.
Ideally, both flows are tested, but that adds overhead to the CI. What would be your recommendation?
1. only test the default flow, where the libarary is included in the native-image and unpacked at run-time
2. only test the configured flow, where the library is unpacked at build-time and is present next to the native-image at run-time
3. test both cases in the CI

Reply:
we should test both cases in the CI

Reply:
Turns out this is still somewhat an issue.
The second case, where the library is unpacked and placed next to the native-image does not work out of the box _yet_ on unix systems. It will work in the next version because of this change: https://github.com/oracle/graal/pull/5932
However it does already work if you add `-Djava.library.path=<path to target folder>` to the runtime arguments.
On windows it does already work in the current version as it already takes the dir of the executable into account.

So to reproduce, when running `mvn verify -Pnative` on a unix system, there will be a `native-tests` executable left in the target directory.
So `./native-tests` will run the tests, and replicate the failures the CI is currently seeing.
`./native-tests -Djava.library.path=.` will run all tests successfully.

What is the best approach here. Document this as a current limitation, to be resolved in GraalVM 23.0? Or wait for the release and set 23.0 as the minimal supported version?

Reply:
we should have 2 different CI job to test each scenario, for each of the 3 OSes. If i understand correctly, for now on Linux (and Mac?) it would fail. We can comment that CI job for now, indicating that it should be solved with the new version.